From d1a05b081b68caa994ccb268ba939662f9534805 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Fri, 15 Jul 2011 21:49:50 +0300
Subject: [PATCH 32/32] (backport) P2P: Do not reply to Probe Request frame indicating only 802.11b rates

Per P2P specification 2.4.1, P2P Device shall shall not respond to
Probe Request frames that indicate support for only 802.11b rates.
---
 src/p2p/p2p.c |   29 +++++++++++++++++++++++++++++
 1 files changed, 29 insertions(+), 0 deletions(-)

diff --git a/src/p2p/p2p.c b/src/p2p/p2p.c
index 016a121..0ea139b 100644
--- a/src/p2p/p2p.c
+++ b/src/p2p/p2p.c
@@ -1641,6 +1641,30 @@ struct wpabuf * p2p_build_probe_resp_ies(struct p2p_data *p2p)
 }
 
 
+static int is_11b(u8 rate)
+{
+	return rate == 0x02 || rate == 0x04 || rate == 0x0b || rate == 0x16;
+}
+
+
+static int supp_rates_11b_only(struct ieee802_11_elems *elems)
+{
+	int num_11b = 0, num_others = 0;
+	int i;
+
+	if (elems->supp_rates == NULL && elems->ext_supp_rates == NULL)
+		return 0;
+
+	for (i = 0; elems->supp_rates && i < elems->supp_rates_len; i++) {
+		if (is_11b(elems->supp_rates[i]))
+			num_11b++;
+		else
+			num_others++;
+	}
+
+	return num_11b > 0 && num_others == 0;
+}
+
 static void p2p_reply_probe(struct p2p_data *p2p, const u8 *addr, const u8 *ie,
 			    size_t ie_len)
 {
@@ -1673,6 +1697,11 @@ static void p2p_reply_probe(struct p2p_data *p2p, const u8 *addr, const u8 *ie,
 		return;
 	}
 
+	if (supp_rates_11b_only(&elems)) {
+		/* Indicates support for 11b rates only */
+		return;
+	}
+
 	/* Check Requested Device Type match */
 	wps = ieee802_11_vendor_ie_concat(ie, ie_len, WPS_DEV_OUI_WFA);
 	if (wps && !p2p_match_dev_type(p2p, wps)) {
-- 
1.7.0.4

